util/util_tls.go

```diff
@@ -0,0 +1,171 @@
+		Subject:            pkix.Name{Organization: []string{orgName}, CommonName: host},
+		NotBefore:          time.Now(),
+		NotAfter:           time.Now().AddDate(1, 0, 0),
+		SignatureAlgorithm: x509.SHA256WithRSA,
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	if ip := net.ParseIP(host); ip != nil {
+		ctx.WithField("ip", ip).Debug("ip used in cert template")
+		tmpl.IPAddresses = append(tmpl.IPAddresses, ip)
+	} else {
+		ctx.WithField("host", host).Debug("host used in cert template")
+		tmpl.DNSNames = append(tmpl.DNSNames, host)
+	}
+
+	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+
+	// gen cert file
+	ctx.WithField("certFile", certPath).Debug("creating cert file")
+	certBlock, err := x509.CreateCertificate(rand.Reader, &tmpl, &tmpl, &privKey.PublicKey, privKey)
+	if err != nil {
+		return err
+	}
+
+	certFile, err := os.Create(certPath)
+	if err != nil {
+		return err
+	}
+	defer certFile.Close()
+	if err := pem.Encode(certFile, &pem.Block{Type: certBlockType, Bytes: certBlock}); err != nil {
+		return err
+	}
+
+	// gen key file
+	ctx.WithField("keyFile", keyPath).Debug("creating key file")
+	keyFile, err := os.OpenFile(keyPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return err
+	}
+	defer keyFile.Close()
+
+	keyBlock := x509.MarshalPKCS1PrivateKey(privKey)
+	if err != nil {
+		return err
+	}
+
+	if err := pem.Encode(keyFile, &pem.Block{Type: keyBlockType, Bytes: keyBlock}); err != nil {
+		return err
+	}
+
+	ctx.WithFields(log.Fields{
+		"certPath": certPath,
+		"certKey":  certPath,
+	}).Debug("self-cert files created")
+
+	return nil
+}
+
+func AssertTrustedHost(ctx apitypes.Context, host, algo string, fingerprint []byte) bool {
+	trusted := "no"
+	fmt.Printf("Rejecting connection to unknown host %s.\n", host)
+	fmt.Printf("SHA Fingerprint presented: %s:%x.\n", algo, fingerprint[0:32])
+	fmt.Print("Do you want to save host to knwon_hosts file? (yes/no): ")
+	fmt.Scanf("%s", &trusted)
+	if strings.ToLower(trusted) == "yes" {
+		return true
+	}
+	return false
+}
+
+func AddKnownHost(ctx apitypes.Context, knownHostPath, host, algo string, fingerprint []byte) error {
+	knownHostPathDir := filepath.Dir(knownHostPath)
+	_, err := os.Stat(knownHostPathDir)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+		if err := os.MkdirAll(knownHostPathDir, 0755); err != nil {
+			return err
+		}
+		ctx.WithField("dir", knownHostPathDir).Debug("created directory for known_hosts")
+	}
+
+	khFile, err := os.OpenFile(knownHostPath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0600)
+	if err != nil {
+		return err
+	}
+	defer khFile.Close()
+
+	fmt.Fprintf(khFile, "%s %s %x\n", host, algo, fingerprint)
+
+	ctx.WithFields(log.Fields{
+		"host": host, "algo": algo, "fingerprint": fmt.Sprintf("%x", fingerprint)},
```

Comment:
@akutz that strict 80-char rule, is that a Govet or Golint rule ?  In the age of 4K HD displays, limiting code to only 80 characters seems to force unnecessary wraps.  Specially strict-80.  I think between 80 to 120 works.